lib/portlayer/exec/base.go

```diff
@@ -0,0 +1,290 @@
+	key := fmt.Sprintf("guestinfo.vice..sessions|%s.started", c.ExecConfig.ID)
+	var detail string
+
+	// Wait some before giving up...
+	ctx, cancel := context.WithTimeout(ctx, propertyCollectorTimeout)
+	defer cancel()
+
+	detail, err = c.vm.WaitForKeyInExtraConfig(ctx, key)
+	if err != nil {
+		return fmt.Errorf("unable to wait for process launch status: %s", err.Error())
+	}
+
+	if detail != "true" {
+		return errors.New(detail)
+	}
+
+	return nil
+}
+
+func (c *containerBase) stop(ctx context.Context, waitTime *int32) error {
+	// make sure we have vm
+	if c.vm == nil {
+		return NotYetExistError{c.ExecConfig.ID}
+	}
+
+	// get existing state and set to stopping
+	// if there's a failure we'll revert to existing
+
+	err := c.shutdown(ctx, waitTime)
+	if err == nil {
+		return nil
+	}
+
+	log.Warnf("stopping %s via hard power off due to: %s", c.ExecConfig.ID, err)
+
+	_, err = tasks.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
+		return c.vm.PowerOff(ctx)
+	})
+
+	if err != nil {
+
+		// It is possible the VM has finally shutdown in between, ignore the error in that case
+		if terr, ok := err.(task.Error); ok {
+			switch terr := terr.Fault().(type) {
+			case *types.InvalidPowerState:
+				if terr.ExistingState == types.VirtualMachinePowerStatePoweredOff {
+					log.Warnf("power off %s task skipped (state was already %s)", c.ExecConfig.ID, terr.ExistingState)
+					return nil
+				}
+				log.Warnf("invalid power state during power off: %s", terr.ExistingState)
+
+			case *types.GenericVmConfigFault:
+
+				// Check if the poweroff task was canceled due to a concurrent guest shutdown
+				if len(terr.FaultMessage) > 0 && terr.FaultMessage[0].Key == vmNotSuspendedKey {
+					log.Infof("power off %s task skipped due to guest shutdown", c.ExecConfig.ID)
+					return nil
+				}
+				log.Warnf("generic vm config fault during power off: %#v", terr)
+
+			default:
+				log.Warnf("hard power off failed due to: %#v", terr)
+			}
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func (c *containerBase) shutdown(ctx context.Context, waitTime *int32) error {
+	// make sure we have vm
+	if c.vm == nil {
+		return NotYetExistError{c.ExecConfig.ID}
+	}
+
+	wait := 10 * time.Second // default
+	if waitTime != nil && *waitTime > 0 {
+		wait = time.Duration(*waitTime) * time.Second
+	}
+
+	cs := c.ExecConfig.Sessions[c.ExecConfig.ID]
+	stop := []string{cs.StopSignal, string(ssh.SIGKILL)}
+	if stop[0] == "" {
+		stop[0] = string(ssh.SIGTERM)
+	}
+
+	for _, sig := range stop {
+		msg := fmt.Sprintf("sending kill -%s %s", sig, c.ExecConfig.ID)
+		log.Info(msg)
+
+		err := c.startGuestProgram(ctx, "kill", sig)
+		if err != nil {
+			return fmt.Errorf("%s: %s", msg, err)
+		}
+
+		log.Infof("waiting %s for %s to power off", wait, c.ExecConfig.ID)
+		timeout, err := c.waitForPowerState(ctx, wait, types.VirtualMachinePowerStatePoweredOff)
+		if err == nil {
+			return nil // VM has powered off
+		}
+
+		if !timeout {
```

Comment:
we're looping over a range of possible signals and trying each - if we return on !err then we don't try SIGKILL